
# Version 1.1 - Protein and Mutation Analysis

<h3> Features </h3>

- downloading and processing all protein sequence data from the NCBI database (currently 10 million lines of data)

- organizing and categorizing records based on labeled protein names

- determinining variations within each protein for every record

- using the most common protein sequence as the baseline for charting deviations from the "mean"

- graphing different proteins to determine which proteins are mutating the fastest (or with more variance) in COVID

- implementing the BLOSUM matrix to determine how significant a mutation is




![alt text](https://github.com/knightsUCF/COVIDGenomeAnalysis/blob/main/Images/COVID%20Structure.png)











Structure of SARS-CoV-2

The SARS-CoV-2 virus consists of four structural proteins: 

S (spike)

E (envelope) surrounds the nucleocapsid

M (membrane) surrounds the nucleocapsid

N protein, or nucleocapsid, which holds the viral RNA

Following infection with SARS-CoV-2, the S and N proteins are the major antigenic targets for antibodies, (e.g., IgG, IgM, IgA)



https://www.pennmedicine.org/updates/blogs/penn-physician-blog/2020/august/covid19-vaccine-research-at-penn-searching-for-covid-immunity




<br>

---

<br>


# Classes

The two main classes that drive the tools for protein analysis are records.py and protein.py:


<h3> records.py </h3>
    
    
```python

from Bio import SeqIO


# download from: https://www.ncbi.nlm.nih.gov/sars-cov-2/ (protein)


class Records:

    def get_all_protein_sequences(self, file_path):
        print('Categorizing protein sequences...\n')
        proteins = {}
        for record in SeqIO.parse(file_path, "fasta"):
            labels = record.description.split()
            if labels[1][1:] not in proteins:
                proteins[labels[1][1:]] = []
            sequence = record.seq
            proteins[labels[1][1:]].append(str(sequence))
        return proteins
```


<h3> protein.py </h3>


```python

from difflib import SequenceMatcher
import operator
import blosum


blosum = blosum.Blosum('blosum.txt')


class Protein:

    def determine_similarity(self, protein_records):
        for protein in protein_records:
            compare_to = protein_records[protein][0] # compare everything to the first protein record, can also compare to a selected sequence
            print('Starting similarity (mutation) analysis for protein: ', protein)
            for record in protein_records[protein]:
                how_similar = SequenceMatcher(None, compare_to, record).ratio()
                print(how_similar)
                
              
    def determine_similarity_by_protein(self, protein_records, protein_name):
        compare_to = protein_records[protein_name][0]
        for record in protein_records[protein_name]:
            how_similar = SequenceMatcher(None, compare_to, record).ratio()
            print(how_similar, ' - ', record)
                
                
    def get_all_labeled_protein_names(self, protein_records):
        protein_names = []
        for protein in protein_records:
            protein_names.append(protein)
        return protein_names


    def determine_frequent_protein_formulas(self, protein_records, protein_name):
        common_protein_formulas = {}
        for record in protein_records[protein_name]:
            if record not in common_protein_formulas:
                common_protein_formulas[record] = 1
            else:
                common_protein_formulas[record] += 1
        common_protein_formulas = dict(sorted(common_protein_formulas.items(), key=operator.itemgetter(1),reverse=True))
        return common_protein_formulas


    # for getting the key with the highest occuring frequency as the baseline to assess mutations from
    # input: protein.DetermineFrequentProteinFormulas(protein_records, protein_name)
    def get_key_with_highest_value(self, dictionary):
        return max(dictionary, key=dictionary.get)


    def calculate_blosum_matrix_on_sequences(self, baseline_sequence, sequences):
        data = {} # returning data for graphing, x axis: frequency occurrence, y axis: BLOSUM rating
        print('Running BLOSUM matrix')
        total_blosum_score = 0
        current_sequence_id = 0
        for current_key, value in sequences.items():
            current_sequence_id += 1
            # skip the first record which we are comparing to, since this is the most frequent
            # if graphing don't skip this so we have a reference
            # if current_sequence_id == 1:
            #    continue
            # only compare sequences of the same length
            if len(baseline_sequence) != len(current_key):
                continue
            # only test the most common frequencies
            if current_sequence_id > 10:
                break
            try:
                for i in range(len(current_key)):
                    # throw out sequences with unknown 'X'
                    if 'X' in current_key:
                        continue
                    if current_key[i] == baseline_sequence[i]:
                        pass
                    else:
                        # ignore missing 'X' data for both the baseline key and match
                        if current_key[i] != 'X' and baseline_sequence[i] != 'X':
                            # print('no match: ', current_key[i], baseline_sequence[i])
                            total_blosum_score += float(blosum.lookup(current_key[i], baseline_sequence[i]))
                print('\nanalysis complete for sequence ID: ', current_sequence_id)
                print('sequence: ', current_key)
                print('sequence frequency out of all records: ', value)
                print('blosum matrix score: ', total_blosum_score, '\n')
                data[value] = total_blosum_score
                # reset
                total_blosum_score = 0
            except Exception as e:
                print(e)
        return data
```



<h3> Organizing Protein Sequences from All Available Records from NCBI (10 million lines) </h3>

The output of GetAllProteinSequences() is a dictionary container with the protein name for the key, and the protein variation formula for the values.

```python


class Records:

    def get_all_protein_sequences(self, file_path):
        print('Categorizing protein sequences...\n')
        proteins = {}
        for record in SeqIO.parse(file_path, "fasta"):
            labels = record.description.split()
            if labels[1][1:] not in proteins:
                proteins[labels[1][1:]] = []
            sequence = record.seq
            proteins[labels[1][1:]].append(str(sequence))
        return proteins
 ```



<h3> Get All Labeled Protein Names from Records </h3>

By getting all the labeled protein names by the original collectors we can later do operations specific to the protein. 

```python
import records
import protein


records = records.Records()
protein = protein.Protein()


file_path = 'sequences.fasta'
   
protein_records = records.get_all_protein_sequences(file_path)

protein.get_all_labeled_protein_names(protein_records)


'''
output

Categorizing protein sequences...


nsp2
nsp3
nsp4
3C-like
nsp6
nsp7
nsp8
nsp9
nsp10
ORF7b
ORF1a
RNA-dependent
helicase
3'-to-5'
endoRNAse
2'-O-ribose
nsp11
ORF10
ORF1ab
surface
ORF3a
envelope
membrane
ORF6
ORF7a
ORF8
nucleocapsid
ORf6
N
ORF7b,
truncated
Chain
E
ORF7a/ORF7b
ORF1ap
orf1ab
orf3a
orf6
orf7a
orf8
orf10
orf1a
RecName:
S
M
ORF3a,
spike
nonstructural
structural
matrix
ORF3
ORF7
'''
```

<h3> Determining Protein Similarity (Mutation) by Protein Name </h3>

Determining protein similarities for different proteins and records can take a long time. Once we run:

    protein.get_all_labeled_protein_names(protein_records)
    



We can then select a similarity analysis by protein name: ORF3a


"The greatest reduction in virus growth was noted following ORF3a deletion."

https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1287583/


ORF3a mutations:

https://github.com/knightsUCF/COVIDGenomeAnalysis/blob/main/results/ORF3a_mutations.txt



```python

import records
import protein


records = records.Records()
protein = protein.Protein()


file_path = 'sequences.fasta'

protein_records = records.get_all_protein_sequences(file_path)

# select any of the protein labels extracted from records above: 
protein_name = 'ORF3a'

protein.determine_similarity_by_rotein(protein_records, protein_name)



'''
output:

ORF3a mutations:

https://github.com/knightsUCF/COVIDGenomeAnalysis/blob/main/results/ORF3a_mutations.txt

(comparing to the first protein instance)
# todo: specify custom protein sequence,
# or compare to the most common found one

1.0 = 100% similar


1.0
0.20363636363636364
0.20363636363636364
0.20363636363636364
0.20363636363636364
0.20363636363636364
0.02909090909090909
0.7890909090909091
0.20363636363636364
0.20363636363636364
0.92
1.0
0.20363636363636364
1.0

...
'''

```




<h3> Comparing Protein Mutations </h3>


```python

import records
import protein

records = records.Records()
protein = protein.Protein()



# click: download -> download all protein sequences: https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus?SeqType_s=Nucleotide&VirusLineage_ss=SARS-CoV-2,%20taxid:2697049
# replace file path with file path of download

file_path = 'sequences.fasta'
protein_records = records.get_all_protein_sequences(file_path)

# this can take a while
# note: protein names are organized by how the original uploader of the record filed them

protein.determine_similarity(protein_records)
```


# Rank Frequency of Protein Formulas by Selected Protein



For a specific protein we can get the most common protein formulas ranked by frequency in the database. The most frequent specific letter for letter variation of the ORF3a protein sequence has appeared nearly 20,000 times out of the entire available records at NCBI!


![alt text](https://github.com/knightsUCF/COVIDGenomeAnalysis/blob/main/Images/ORF3a.png)



```python

import records
import protein


records = records.Records()
protein = protein.Protein()


file_path = 'sequences.fasta'

protein_records = records.get_all_protein_sequences(file_path)


# select any of the protein labels extracted from records above

protein_name = 'ORF3a'

frequent_protein_sequences = protein.determine_frequent_protein_formulas(protein_records, protein_name)


# output 

for key, value in frequent_protein_sequences.items():
    print(key, ' - ', value, '\n')


'''
available protein labels:

nsp2
nsp3
nsp4
3C-like
nsp6
nsp7
nsp8
nsp9
nsp10
ORF7b
ORF1a
RNA-dependent
helicase
3'-to-5'
endoRNAse
2'-O-ribose
nsp11
ORF10
ORF1ab
surface
ORF3a
envelope
membrane
ORF6
ORF7a
ORF8
nucleocapsid
ORf6
N
ORF7b,
truncated
Chain
E
ORF7a/ORF7b
ORF1ap
orf1ab
orf3a
orf6
orf7a
orf8
orf10
orf1a
RecName:
S
M
ORF3a,
spike
nonstructural
structural
matrix
ORF3
ORF7


output

https://github.com/knightsUCF/COVIDGenomeAnalysis/blob/main/results/ORF3A_Formula_Frequencies.txt

(scroll all the way right to get the frequency count of each protein formula chain)

...


MDLFMR ... TSVPL  -  93 

MDLFM ... TSVPL  -  96 

MDLFMR ... YDEPTTTTSVPL  -  310 

MDLFM ... IYDEPTTTTSVPL  -  374 

MDLFMR... MEPIYDEPTTTTSVPL  -  530 

MDL ... TSVPL  -  11495 

MDLFM ... EPIYDEPTTTTSVPL  -  19805 



'''

```
    

# Protein Mutation Analysis

The goals for COVID protein mutation analysis:

- take the most common protein as the baseline

- graph the deviations from this baseline

- graph mutation deviations for different proteins, and determine which ones are mutating the fastest
    


# Assessing Mutations with the Blosum Matrix

"Substituting an amino acid with another from the same category is more likely to have a smaller impact on the structure and function of a protein than replacement with an amino acid from a different category."

https://en.wikipedia.org/wiki/BLOSUM


Here we will only determine different amino acid substitions. If we were substiting the same amino acid but with a different codon formula we would have analyze the nucleotide sequence to determine synonymous codons, something we can do later. Substituting the same amino acid but with a different codon formula also seems to have less significance in the BLOSUM matrix. 



 ```python
 
import records
import protein
import operator
import blosum


records = records.Records()
protein = protein.Protein()
blosum = blosum.Blosum('blosum.txt')


file_path = 'sequences.fasta'

protein_records = records.get_all_protein_sequences(file_path)

protein_name = 'ORF3a'


frequent_protein_sequences = protein.determine_frequent_protein_formulas(protein_records, protein_name)


# for getting the key with the highest occuring frequency as the baseline we will be comparing other mutations to
baseline_protein_sequence = protein.get_key_with_highest_value(frequent_protein_sequences)


# we want to check where these mutations are occuring and apply the Blosum matrix
# this analysis assumes a few factors:
# - currently only analyzing sequences of the same length
# - to test sequences of different lengths a method needs to be established, such as how should regional shifting be treated
# - only comparing the records up to a significant frequency (ignore records of only 1 variation, stop near records with over 100 records in the database)


print('baseline sequence of protein: ', protein_name, ' ', baseline_protein_sequence)
print('baseline protein occuring frequency: ', frequent_protein_sequences[baseline_protein_sequence])


protein.calculate_blosum_matrix_on_sequences(baseline_protein_sequence, frequent_protein_sequences)


# output: https://github.com/knightsUCF/COVIDGenomeAnalysis/blob/main/results/ORF3a_Blosum_Matrix.txt

```


# BLOSUM Matrix Graphs


When graphing the BLOSUM matrix mutations in COVID an interesting feature emerged -- there is a steep dropoff after the most frequent protein formula. Also the next common protein formula variations have a relatively high BLOSUM index (a lower BLOSUM index = more unstable mutation) which could suggest that the most common COVID protein sequences prefer stable configurations.

An alternative explanation could be that the most common COVID strain has a higher rate of transmittance. 



![alt text](https://github.com/knightsUCF/COVIDGenomeAnalysis/blob/main/Images/BLOSUM%20Mutations%20-%20Nucleocapsid.png)


![alt text](https://github.com/knightsUCF/COVIDGenomeAnalysis/blob/main/Images/BLOSUM%20Mutations%20-%20ORF1a.png)


![alt text](https://github.com/knightsUCF/COVIDGenomeAnalysis/blob/main/Images/BLOSUM%20Mutations%20-%20ORF1ab.png)


![alt text](https://github.com/knightsUCF/COVIDGenomeAnalysis/blob/main/Images/BLOSUM%20Mutations%20-%20ORF7b.png)


![alt text](https://github.com/knightsUCF/COVIDGenomeAnalysis/blob/main/Images/BLOSUM%20Mutations%20-%20ORF8.png)




```python
import records
import protein
import operator
import blosum
import graph


records = records.Records()
protein = protein.Protein()
blosum = blosum.Blosum('blosum.txt')



file_path = 'sequences.fasta'
protein_records = records.get_all_protein_sequences(file_path)



protein_name = 'ORF7b' # -500, 40,0000

# protein_name = 'ORF1a' # -1000, 10,000

# protein_name = 'ORF1ab' # -1000, 6,000

# protein_name = 'ORF8' # -1000, 4,000

# protein_name = 'nucleocapsid' # -500, 2000



frequent_protein_sequences = protein.determine_frequent_protein_formulas(protein_records, protein_name)


# for getting the key with the highest occuring frequency as the baseline to assess mutations from
# this is the the baseline we will be comparing other mutations to
baseline_protein_sequence = protein.get_key_with_highest_value(frequent_protein_sequences)


# we want to check where these mutations are occuring and apply the Blosum matrix
# this analysis assumes a few factors:
# - currently only analyzing sequences of the same length
# - since we are testing sequences of the same length
# - to test sequences of different lengths a method needs to be established, such as how should regional shifting be treated
# - only comparing the records up to a significant frequency
# - nevertheless we can still compare how this analysis stacks up between the different proteins

print('baseline sequence of protein: ', protein_name, ' ', baseline_protein_sequence)
print('baseline protein occuring frequency: ', frequent_protein_sequences[baseline_protein_sequence])


data = protein.calculate_blosum_matrix_on_sequences(baseline_protein_sequence, frequent_protein_sequences)



x_set = []
y_set = []



for i in data:
    x_set.append(i)
    y_set.append(data[i])


title = 'BLOSUM Mutations - ORF8'

x_label = 'Frequency Occurrence'
y_label = 'BLOSUM Rating'

x_start = -1000
x_end = 4000

y_start = -5
y_end = 5


graph = graph.Graph(title, x_label, y_label, x_start, x_end, y_start, y_end)

graph.add_data(x_set, y_set, 'green')



graph.show()

```


# A Note On Comparing Different Lengths Sequences

With the above method we are only comparing identical lengths of proteins formulas from different samples. How should we treat different lengths of proteins? "The nontrivial answer to this question depends entirely on the particular structure or functional aspect one wants to investigate. The necessary and sufficient similarity threshodlw ill be different for each task... Sander and Schneider pioneered the algorithmci ivnestigation of the relationship between protein sequence similarity and structural similarity." pg 32 of Bioinfomrations - The Machine Learning Approach
