
# Version 1.1 - Protein and Mutation Analysis

<h3> Goals </h3>

- download and process all protein sequence data from the NCBI database (currently 10 million lines of data)

- organize and categorize records based on labeled protein names

- determine variations within each protein for each record

- use the most common protein sequence as the baseline for charting deviations from the "mean"

- graph all proteins to determine which proteins are mutating the fastest (or with more variance) in COVID


![alt text](https://github.com/knightsUCF/COVIDGenomeAnalysis/blob/main/Images/COVID%20Structure.png)

---
<br>

# CONTENTS


I. Downloading all Available Protein Sequences from the NCBI Database

II. Organizing Protein Sequences by Individual Protein

III. Getting All Labeled Protein Names from Records

IV. Determining Protein Similarity (Mutation) by Protein Name

V. Comparing Protein Mutations

VI. Ranking Frequency of Protein Formulas by Selected Protein

![alt text](https://github.com/knightsUCF/COVIDGenomeAnalysis/blob/main/Images/mutations.jpg)

<br>

---

<br>

# FEATURES

- process the entire catalog of downloaded COVID protein sequences from: https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus?SeqType_s=Nucleotide&VirusLineage_ss=SARS-CoV-2,%20taxid:2697049

- in records.py we automatically create an organized list which categorizes each protein variation sequence under the labeled name of the protein


- we can get all available labeled protein names from the database: `protein.GetAllLabeledProteinNames(protein_records)`


- we can get protein similiarities (mutations) by name, since this can save us a lot of time instead of processing all protein mutations `protein.GetAllLabeledProteinNames(protein_records)`

<br>

---

<br>


# Classes

The two main classes that drive the tools for protein analysis are records.py and protein.py:


<h3> records.py </h3>
    
    
```python

from Bio import SeqIO


# download from: https://www.ncbi.nlm.nih.gov/sars-cov-2/ (protein)


class Records():

    def GetAllProteinSequences(self, file_path):
        print('Categorizing protein sequences...\n')
        proteins = {}
        for record in SeqIO.parse(file_path, "fasta"):
            labels = record.description.split()
            if labels[1][1:] not in proteins:
                proteins[labels[1][1:]] = []
            sequence = record.seq
            proteins[labels[1][1:]].append(str(sequence))
        return proteins
```


<h3> protein.py </h3>


```python
from difflib import SequenceMatcher



class Protein():

    def DetermineSimilarity(self, protein_records):
        for protein in protein_records:
            compare_to = protein_records[protein][0] # compare everything to the first protein record, can also compare to a selected sequence
            print('Starting similarity (mutation) analysis for protein: ', protein)
            for record in protein_records[protein]:
                how_similar = SequenceMatcher(None, compare_to, record).ratio()
                print(how_similar)
                
              
    def DetermineSimilarityByProtein(self, protein_records, protein_name):
        compare_to = protein_records[protein_name][0]
        for record in protein_records[protein_name]:
            how_similar = SequenceMatcher(None, compare_to, record).ratio()
            print(how_similar, ' - ', record)
                
                
    def GetAllLabeledProteinNames(self, protein_records):
        protein_names = []
        for protein in protein_records:
            protein_names.append(protein)
        return protein_names


    def DetermineFrequentProteinFormulas(self, protein_records, protein_name):
        common_protein_formulas = {}
        for record in protein_records[protein_name]:
            if record not in common_protein_formulas:
                common_protein_formulas[record] = 1
            else:
                common_protein_formulas[record] += 1
        common_protein_formulas =  dict(sorted(common_protein_formulas.items(), key=lambda item: item[1]))
        return common_protein_formulas
```




<h3> Organizing Protein Sequences from All Available Records from NCBI (10 million lines) </h3>

The output of GetAllProteinSequences() is a dictionary container with the protein name for the key, and the protein variation formula for the values.

```python


class Records():

    def GetAllProteinSequences(self, file_path):
        print('Categorizing protein sequences...\n')
        proteins = {}
        for record in SeqIO.parse(file_path, "fasta"):
            labels = record.description.split()
            if labels[1][1:] not in proteins:
                proteins[labels[1][1:]] = []
            sequence = record.seq
            proteins[labels[1][1:]].append(str(sequence))
        return proteins
 ```



<h3> Get All Labeled Protein Names from Records </h3>

By getting all the labeled protein names by the original collectors we can later do operations specific to the protein. 

```python
import records
import protein


records = records.Records()
protein = protein.Protein()


file_path = 'sequences.fasta'
   
protein_records = records.GetAllProteinSequences(file_path)

protein.GetAllLabeledProteinNames(protein_records)


'''
output

Categorizing protein sequences...


nsp2
nsp3
nsp4
3C-like
nsp6
nsp7
nsp8
nsp9
nsp10
ORF7b
ORF1a
RNA-dependent
helicase
3'-to-5'
endoRNAse
2'-O-ribose
nsp11
ORF10
ORF1ab
surface
ORF3a
envelope
membrane
ORF6
ORF7a
ORF8
nucleocapsid
ORf6
N
ORF7b,
truncated
Chain
E
ORF7a/ORF7b
ORF1ap
orf1ab
orf3a
orf6
orf7a
orf8
orf10
orf1a
RecName:
S
M
ORF3a,
spike
nonstructural
structural
matrix
ORF3
ORF7
'''
```

<h3> Determining Protein Similarity (Mutation) by Protein Name </h3>

Determining protein similarities for all proteins and all records can take a long time. Once we run:

    protein.GetAllLabeledProteinNames(protein_records)
    



We can then select a similarity analysis by protein name: ORF3a


"The greatest reduction in virus growth was noted following ORF3a deletion."

https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1287583/


ORF3a mutations:

https://github.com/knightsUCF/COVIDGenomeAnalysis/blob/main/results/ORF3a_mutations.txt




    import records
    import protein
    
    
    records = records.Records()
    protein = protein.Protein()


    file_path = 'sequences.fasta'
    
    protein_records = records.GetAllProteinSequences(file_path)
    
    # select any of the protein labels extracted from records above: 
    protein_name = 'ORF3a'
    
    protein.DetermineSimilarityByProtein(protein_records, protein_name)



    '''
    output:
    
    ORF3a mutations:
    
    https://github.com/knightsUCF/COVIDGenomeAnalysis/blob/main/results/ORF3a_mutations.txt
    
    (comparing to the first protein instance)
    # todo: specify custom protein sequence,
    # or compare to the most common found one
    
    1.0 = 100% similar
    

    1.0
    0.20363636363636364
    0.20363636363636364
    0.20363636363636364
    0.20363636363636364
    0.20363636363636364
    0.02909090909090909
    0.7890909090909091
    0.20363636363636364
    0.20363636363636364
    0.92
    1.0
    0.20363636363636364
    1.0
    
    ...
    '''






<h3> Comparing Protein Mutations </h3>


    import records
    import protein
    
    records = records.Records()
    protein = protein.Protein()
    
    

    # click: download -> download all protein sequences: https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus?SeqType_s=Nucleotide&VirusLineage_ss=SARS-CoV-2,%20taxid:2697049
    # replace file path with file path of download
    
    file_path = 'sequences.fasta'
    protein_records = records.GetAllProteinSequences(file_path)

    # this can take a while
    # note: protein names are organized by how the original uploader of the record filed them
    
    protein.DetermineSimilarity(protein_records)



# Rank Frequency of Protein Formulas by Selected Protein



For a specific protein we can get the most common protein formulas ranked by frequency in the database. The most frequent specific letter for letter variation of the ORF3a protein sequence has appeared nearly 20,000 times out of the entire available records at NCBI!


![alt text](https://github.com/knightsUCF/COVIDGenomeAnalysis/blob/main/Images/ORF3a.png)



```python

import records
import protein


records = records.Records()
protein = protein.Protein()


file_path = 'sequences.fasta'

protein_records = records.GetAllProteinSequences(file_path)


# select any of the protein labels extracted from records above

protein_name = 'ORF3a'

frequent_protein_sequences = protein.DetermineFrequentProteinFormulas(protein_records, protein_name)


# output 

for key, value in frequent_protein_sequences.items():
    print(key, ' - ', value, '\n')


'''
available protein labels:

nsp2
nsp3
nsp4
3C-like
nsp6
nsp7
nsp8
nsp9
nsp10
ORF7b
ORF1a
RNA-dependent
helicase
3'-to-5'
endoRNAse
2'-O-ribose
nsp11
ORF10
ORF1ab
surface
ORF3a
envelope
membrane
ORF6
ORF7a
ORF8
nucleocapsid
ORf6
N
ORF7b,
truncated
Chain
E
ORF7a/ORF7b
ORF1ap
orf1ab
orf3a
orf6
orf7a
orf8
orf10
orf1a
RecName:
S
M
ORF3a,
spike
nonstructural
structural
matrix
ORF3
ORF7


output

https://github.com/knightsUCF/COVIDGenomeAnalysis/blob/main/results/ORF3A_Formula_Frequencies.txt

(scroll all the way right to get the frequency count of each protein formula chain)

...


MDLFMR ... TSVPL  -  93 

MDLFM ... TSVPL  -  96 

MDLFMR ... YDEPTTTTSVPL  -  310 

MDLFM ... IYDEPTTTTSVPL  -  374 

MDLFMR... MEPIYDEPTTTTSVPL  -  530 

MDL ... TSVPL  -  11495 

MDLFM ... EPIYDEPTTTTSVPL  -  19805 



'''

```
    

# Protein Mutation Analysis

The goals for COVID protein mutation analysis:

- take the most common protein as the baseline

- graph the percent deviations from this baseline

- graph mutation deviations for all the major proteins, to determine which ones are mutating the fastest
    




# TODO:

- get all labeled protein names

- do similarity analysis for specific proteins

- graph mutations
